File: boidCollection.js
// Haakon8855

class BoidCollection {
    constructor(numBoids) {
        this.numBoids = numBoids;
        this.boids = [];

        for (let i = 0; i < this.numBoids; i++) {
            this.boids.push(
                new Boid(
                    Math.random() * window.innerWidth,
                    Math.random() * window.innerHeight
                )
            );
        }
    }

    move() {
        let viewDist = this.boids[0].getAttribute("viewDist");
        let positions = [];
        let headings = [];
        for (const boid of this.boids) {
            positions.push(boid.position);
            headings.push(boid.heading);
        }

        let visiblePositions = [];
        let visibleHeadings = [];
        for (const boid of this.boids) {
            visiblePositions = [];
            visibleHeadings = [];
            for (let j = 0; j < this.boids.length; j++) {
                if (boid.distanceTo(this.boids[j]) < viewDist) {
                    visiblePositions.push(positions[j]);
                    visibleHeadings.push(headings[j]);
                }
            }
            boid.move(visiblePositions, visibleHeadings);
        }
    }

    getPositions() {
        let positions = [];
        for (const boid of this.boids) {
            positions.push(boid.position);
        }
        return positions;
    }

    getAngles() {
        let angles = [];
        for (const boid of this.boids) {
            angles.push(boid.getAngle());
        }
        return angles;
    }
}

File: Boids/BoidCollection.cs
using Boids.Models.Config;

namespace Boids;

public class BoidCollection
{
    private readonly Config _config;
    private Boid[] Boids { get; }

    public double[][] Positions => Boids.Select(b => b.Position).ToArray();
    public double[][] Headings => Boids.Select(b => b.Heading).ToArray();

    public BoidCollection(Config config)
    {
        _config = config;
        Boids = new Boid[_config.Boids.NumberOfBoids];
        var rand = new Random();

        for (var i = 0; i < _config.Boids.NumberOfBoids; i++)
        {
            Boids[i] = new Boid(
                rand.Next(_config.Window.Width),
                rand.Next(_config.Window.Height),
                _config
            );
        }
    }

    public void Move()
    {
        var viewDist = _config.Boids.ViewDist;
        List<double[]> positions = [];
        List<double[]> headings = [];
        foreach (var boid in Boids)
        {
            positions.Add(boid.Position);
            headings.Add(boid.Heading);
        }

        foreach (var boid in Boids)
        {
            List<double[]> visiblePositions = [];
            List<double[]> visibleHeadings = [];
            for (var j = 0; j < Boids.Length; j++)
            {
                if (boid.DistanceTo(Boids[j]) < viewDist)
                {
                    visiblePositions.Add(positions[j]);
                    visibleHeadings.Add(headings[j]);
                }
            }

            boid.Move(visiblePositions, visibleHeadings);
        }
    }
}